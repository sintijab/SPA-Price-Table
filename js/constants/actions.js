export const SHOW_OVERLAY = 'SHOW_OVERLAY';
export const CLOSE_OVERLAY = 'CLOSE_OVERLAY';

export const DATA_LOADED = 'DATA_LOADED';
export const DATA_FAILED = 'DATA_FAILED';
export const DATA_EMPTY = 'DATA_EMPTY';
export const DATA_UPDATED = 'DATA_UPDATED';
