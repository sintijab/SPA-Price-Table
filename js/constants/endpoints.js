export const ALL_COMPANIES = 'company';
export const BUDGET = 'budget';
export const URL_HOST = window.location.href;
