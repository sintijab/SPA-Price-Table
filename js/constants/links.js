export const MOCKED_URL = 'data.json';
