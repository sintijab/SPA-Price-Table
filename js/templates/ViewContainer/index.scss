@import '../../../styles/main';

.section {
  display: block;
  margin: 2rem;
}

.list_item {
  padding: 1.5rem 2rem;
  border-bottom: 1px solid $light-gray;
  cursor: pointer;
}

.header, .row {
  display: flex;
  line-height: 2;
  padding: 1rem 0;
  border-bottom: 1px solid $light-gray;
  flex-wrap: wrap;
}

.row {
  cursor: pointer;
}

.col {
  flex: 1;
  flex-basis: 25%;

  @include desktop {
	   flex-basis: 20%;
	}
}
