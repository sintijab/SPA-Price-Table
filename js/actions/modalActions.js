import { SHOW_OVERLAY, CLOSE_OVERLAY } from '../constants/actions';

export const openModal = item => (dispatch) => {
  dispatch({
    type: SHOW_OVERLAY,
    item
  });
}

export const closeModal = () => (dispatch) => {
  dispatch({
    type: CLOSE_OVERLAY
  });
}
