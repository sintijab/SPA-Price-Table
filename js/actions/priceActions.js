import { DATA_UPDATED, DATA_FAILED } from '../constants/actions';
import { updatePrice } from '../utils/priceFunctions';

export const updateTotalBudget = (id, price) => (dispatch) => {
  try {
    const newPrice = JSON.parse(price);
    const data = {
      idCompany: id,
      newPrice,
    };
    dispatch({
      type: DATA_UPDATED,
      payload: data
    });
  } catch (err) {
    console.error(err);
    dispatch({
      type: DATA_FAILED,
    });
  }
}
