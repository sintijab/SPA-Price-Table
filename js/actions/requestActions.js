import { DATA_LOADED, DATA_FAILED, DATA_EMPTY } from '../constants/actions';

export const dataLoaded = (data) => (dispatch) => {
  dispatch({
    type: DATA_LOADED,
    response: data
  });
}

export const dataFailed = () => (dispatch) => {
  dispatch({
    type: DATA_FAILED,
  });
}

export const dataEmpty = () => (dispatch) => {
  dispatch({
    type: DATA_EMPTY,
  });
}
