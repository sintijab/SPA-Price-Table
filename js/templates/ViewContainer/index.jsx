/* eslint react/no-danger: 0 */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import DOMPurify from 'dompurify';

import { openModal } from '../../actions/modalActions';
import {
  COMPANY_TITLE,
  COMPANY_BUDGET,
  COMPANY_BUDGET_SPENT,
  COMPANY_BUDGET_LEFT,
  PURCHASE_DATE,
} from '../../constants/titles';
import { useWindowDimensions } from '../../utils/functions';
import {
  validatePrice,
  getActiveBudget,
  displayFormat,
} from '../../utils/priceFunctions';
import styles from './index.scss';

const { sanitize } = DOMPurify;

const ViewContainer = (
  {
    data,
    showModal,
  },
) => {
  const { isMobile, isTablet } = useWindowDimensions();
  const hasSmallViewPort = isMobile || isTablet;
  if (data && data[0] && data[0].data) {
    const rows = data[0].data.map((item) => {
      const fullPrice = validatePrice(item.budget);
      const expenses = validatePrice(item.budget_spent);
      const totalPrice = getActiveBudget(fullPrice, expenses);
      const {
        0: budgetTotal,
        1: budgetSpent,
        2: budgetLeft,
      } = displayFormat([fullPrice, expenses, totalPrice]);
      const {
        name,
        date_of_first_purchase: purchaseDate,
      } = item || {};
      const shortTitle = `<div><div>${name}</div> <div>${purchaseDate}</div></div>`;

      return (
        <div
          key={item.id}
          role="presentation"
          onClick={() => showModal(item)}
          className={styles.row}
        >
          {!hasSmallViewPort && (
            <div
              className={styles.col}
              dangerouslySetInnerHTML={{ __html: sanitize(name) }}
            />
          )}
          {!hasSmallViewPort && (
            <div
              className={styles.col}
              dangerouslySetInnerHTML={{ __html: sanitize(purchaseDate) }}
            />
          )}
          {hasSmallViewPort && (
            <div
              className={styles.col}
              dangerouslySetInnerHTML={{ __html: sanitize(shortTitle) }}
            />
          )}
          <div
            className={styles.col}
            dangerouslySetInnerHTML={{ __html: sanitize(budgetTotal) }}
          />
          <div
            className={styles.col}
            dangerouslySetInnerHTML={{ __html: sanitize(budgetSpent) }}
          />
          <div
            className={styles.col}
            dangerouslySetInnerHTML={{ __html: sanitize(budgetLeft) }}
          />
        </div>
      );
    });
    return (
      <section className={styles.section}>
        <header className={styles.header}>
          <div
            className={styles.col}
            dangerouslySetInnerHTML={{ __html: DOMPurify.sanitize(COMPANY_TITLE) }}
          />
          {!hasSmallViewPort && (
            <div
              className={styles.col}
              dangerouslySetInnerHTML={{ __html: DOMPurify.sanitize(PURCHASE_DATE) }}
            />
          )}
          <div
            className={styles.col}
            dangerouslySetInnerHTML={{ __html: DOMPurify.sanitize(COMPANY_BUDGET) }}
          />
          <div
            className={styles.col}
            dangerouslySetInnerHTML={{ __html: DOMPurify.sanitize(COMPANY_BUDGET_SPENT) }}
          />
          <div
            className={styles.col}
            dangerouslySetInnerHTML={{ __html: DOMPurify.sanitize(COMPANY_BUDGET_LEFT) }}
          />
        </header>
        <div className={styles.list_items}>
          {rows}
        </div>
      </section>
    );
  }
  return <div />;
};

const mapStateToProps = (state) => ({
  data: state.requests,
  priceUpdate: state.prices,
});
const mapDispatchToProps = (dispatch) => ({
  showModal: (id) => dispatch(openModal(id)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ViewContainer);

ViewContainer.propTypes = {
  data: PropTypes.arrayOf(PropTypes.object),
  showModal: PropTypes.func,
};

ViewContainer.defaultProps = {
  data: [],
  showModal: () => {},
};
