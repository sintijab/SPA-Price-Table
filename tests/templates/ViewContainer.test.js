import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import thunk from "redux-thunk";
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store'
import Provider from 'react-redux';
import ViewContainer from '../../js/templates/ViewContainer';
import { DATA_LOADED } from '../../js/constants/actions';

Enzyme.configure({ adapter: new Adapter() });
const mockStore = configureMockStore([thunk]);

describe('<ViewContainer />', () => {
  let store;

  beforeEach(() => {
    const data = [{
      id: 7,
      name: "InsuranceFirma",
      budget: 9787866,
      budget_spent: 46000,
      date_of_first_purchase: "2409-07-07"
    }];
      const initialState = {
        modals: [],
        requests: [{
         data,
         type: DATA_LOADED,
         loading: false,
         error: false,
        }],
        prices: [],
      };
      store = mockStore(initialState);
  });

    it('should render ViewContainer container', () => {
      const wrapper = shallow(<ViewContainer store={store} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render ViewContainer container with stored data', () => {
      const wrapper = mount(<ViewContainer store={store} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
});
