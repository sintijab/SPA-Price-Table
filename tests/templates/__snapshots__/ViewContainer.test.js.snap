// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewContainer /> should render ViewContainer container 1`] = `
<ContextProvider
  value={null}
>
  <ViewContainer
    data={
      Array [
        Object {
          "data": Array [
            Object {
              "budget": 9787866,
              "budget_spent": 46000,
              "date_of_first_purchase": "2409-07-07",
              "id": 7,
              "name": "InsuranceFirma",
            },
          ],
          "error": false,
          "loading": false,
          "type": "DATA_LOADED",
        },
      ]
    }
    priceUpdate={Array []}
    showModal={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;

exports[`<ViewContainer /> should render ViewContainer container with stored data 1`] = `
<Connect(ViewContainer)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ViewContainer
    data={
      Array [
        Object {
          "data": Array [
            Object {
              "budget": 9787866,
              "budget_spent": 46000,
              "date_of_first_purchase": "2409-07-07",
              "id": 7,
              "name": "InsuranceFirma",
            },
          ],
          "error": false,
          "loading": false,
          "type": "DATA_LOADED",
        },
      ]
    }
    priceUpdate={Array []}
    showModal={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <section
      className="section"
    >
      <header
        className="header"
      >
        <div
          className="col"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Company",
            }
          }
        />
        <div
          className="col"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Date of purchase",
            }
          }
        />
        <div
          className="col"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Total budget",
            }
          }
        />
        <div
          className="col"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Budget spent",
            }
          }
        />
        <div
          className="col"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Budget left",
            }
          }
        />
      </header>
      <div
        className="list_items"
      >
        <div
          className="row"
          key="7"
          onClick={[Function]}
          role="presentation"
        >
          <div
            className="col"
            dangerouslySetInnerHTML={
              Object {
                "__html": "InsuranceFirma",
              }
            }
          />
          <div
            className="col"
            dangerouslySetInnerHTML={
              Object {
                "__html": "2409-07-07",
              }
            }
          />
          <div
            className="col"
            dangerouslySetInnerHTML={
              Object {
                "__html": "9.787.866,00 €",
              }
            }
          />
          <div
            className="col"
            dangerouslySetInnerHTML={
              Object {
                "__html": "46.000,00 €",
              }
            }
          />
          <div
            className="col"
            dangerouslySetInnerHTML={
              Object {
                "__html": "9.741.866,00 €",
              }
            }
          />
        </div>
      </div>
    </section>
  </ViewContainer>
</Connect(ViewContainer)>
`;
